### PR TITLE
Phone audio sink works

### PR DESCRIPTION
**Pull request by JerryChocoY**

### CODE REVIEW

File: examples/twilio_stream_example.py [hunk 1]
```diff
@@ -132,12 +107,28 @@ async def on_error(error):
...
+                stream_sid = data["streamSid"]
                 await session.start()
+
             if data["event"] == "media":
                 payload = data["media"]["payload"]
                 chunk = base64.b64decode(payload)
                 await source.write(chunk)
+
+                try:
+                    # TODO: something wrong here. The phone receives merchanic voice
```
Comment (juberti) [inline examples/twilio_stream_example.py:119]: This doesn't seem right to me. You probably want to send the bytes on the web socket when they arrive from the TTS sink, but here you're going to only send them when you receive bytes from the remote side.

File: examples/twilio_stream_example.py [hunk 2]
```diff
@@ -132,12 +107,28 @@ async def on_error(error):
...
+                    chunk = sink.read()
+                    ulaw = audioop.lin2ulaw(chunk, 2)
+                    payload = base64.b64encode(ulaw).decode("utf-8")
+                    media_data = {
+                        "event": "media",
+                        "streamSid": stream_sid,
+                        "media": {"payload": payload},
```
Comment (juberti) [inline examples/twilio_stream_example.py:126]: i'm thinking you'll need to set the media `timestamp` key too:
https://www.twilio.com/docs/voice/twiml/stream#message-media

File: examples/twilio_stream_example.py [hunk 3]
```diff
@@ -17,44 +17,18 @@
 class PhoneAudioSink(audio_base.AudioSink):
     """AudioSink that plays to the phone stream."""
 
-    def __init__(self, ws: aiohttp.web.WebSocketResponse) -> None:
+    def __init__(self) -> None:
         super().__init__()
-        self._ws = ws
-        self._stream_sid = ""
-        self._sequence_no = 0
```
Comment (juberti) [inline examples/twilio_stream_example.py:24]: per https://www.twilio.com/docs/voice/twiml/stream#message-media, sequence numbers start at 1

File: examples/twilio_stream_example.py (hunk 1 above)
Comment (JerryChocoY) [reply examples/twilio_stream_example.py:121]: The audio is understandable now, but could be better.

Can you explain more between TTS sink and remote?

The PhoneAudioSink cannot write to WS directly (it only receives the first audio frame). It is OK to write to WS here by reading from the sink queue.

File: examples/twilio_stream_example.py [hunk 4]
```diff
@@ -88,13 +61,28 @@ async def websocket_handler(request):
...
-    sink = PhoneAudioSink(ws)
+    sink = PhoneAudioSink()
     params = VoiceSessionParams(
         agent_id=args.agent,
         tts_voice=args.tts_voice,
     )
     session = VoiceSession(source, sink, params)
 
+    @sink.on("data")
+    async def on_sink_data(data):
```
Comment (JerryChocoY) [inline examples/twilio_stream_example.py:72]: could use some refactor later

**Comment by JerryChocoY** [comment]
the phone audio sink works now.

File: examples/twilio_stream_example.py [hunk 5]
```diff
@@ -27,7 +27,12 @@ async def start(self, sample_rate: int = 8000, num_channels: int = 1):
         pass
 
     async def write(self, chunk: bytes) -> None:
-        self.emit("data", chunk)
+        sample = numpy.frombuffer(chunk, numpy.int16).astype(numpy.float32)
+        resampled = librosa.resample(sample, orig_sr=48000, target_sr=8000).astype(
```
Comment (juberti) [inline examples/twilio_stream_example.py:31]: you'll want to use the `sample_rate` from `start()` here, which indicates the sr of the incoming bytes, as `orig_sr`

File: examples/twilio_stream_example.py [hunk 6]
```diff
@@ -27,7 +27,12 @@ async def start(self, sample_rate: int = 8000, num_channels: int = 1):
         pass
```
Comment (juberti) [inline examples/twilio_stream_example.py:27]: remove the int = 8000 default on line 26, it has no effect here

**Review by juberti — APPROVED** [review]
LG with two minor notes

**Comment by juberti** [comment]
Great progress here. Please make a demo video when you get a chance!

**Review by juberti — APPROVED** [review]

**Pull request merged**